go/vt/vtgate/planbuilder/operators/ast2op.go

```diff
@@ -115,16 +115,205 @@ func createOperatorFromUpdate(ctx *plancontext.PlanningContext, updStmt *sqlpars
+		for _, expr := range updStmt.Exprs {
+			colIdx := fk.ParentColumns.FindColumn(expr.Name.Name)
+			if colIdx < 0 {
+				continue
+			}
+			switch expr.Expr.(type) {
+			case *sqlparser.Argument, *sqlparser.NullVal, sqlparser.BoolVal, *sqlparser.Literal:
+			default:
+				return nil, vterrors.VT12001("foreign keys management at vitess with non-literal values")
+			}
+		}
+
+		// We now construct the update query for the child table -
+		// For CASCADE type constraint, the query looks like this -
+		// 		`UPDATE <child_table> SET <child_column_updated_using_update_exprs_from_parent_update_query> WHERE <child_columns_in_fk> IN (<bind variable for the output from SELECT>)`
+		// For SET NULL type constraint, the query looks like this -
+		//		`UPDATE <child_table> SET <child_column_updated_using_update_exprs_from_parent_update_query>
+		//		WHERE <child_columns_in_fk> IN (<bind variable for the output from SELECT>)
+		//		[AND <child_columns_in_fk> NOT IN (<bind variables in the SET clause of the original update>)]`
+		// 	If all the columns of the foreign key in the parent don't actually end up getting updated, then we shouldn't be setting the child columns to NULL.
+
+		// Some operations are common for both the cases.
+		// 1. building the bind variable name for the output from the SELECT query.
+		bvName := ctx.ReservedVars.ReserveVariable("fkc_vals")
+		// 2. ValTuple having the column names of the foreign key constraint in the child table.
+		var valTuple sqlparser.ValTuple
+		for _, column := range fk.ChildColumns {
+			valTuple = append(valTuple, sqlparser.NewColName(column.String()))
+		}
+		// 3. Use the above 2 to create the comparison expression we use in the where condition of the UPDATE.
+		compExpr := &sqlparser.ComparisonExpr{
+			Operator: sqlparser.InOp,
+			Left:     valTuple,
+			Right:    sqlparser.NewListArg(bvName),
+		}
+
+		// For both the cases, 2 things are different. The where condition and the update expressions.
+		// For CASCADE type constraint, the where condition is the same as the comparison expression above.
+		// For SET NULL type constraint, the where condition is the same as the comparison expression above, with an additional condition (that condition is optional).
+
+		// We create the variables for these 2 things, and populate them in the switch case below.
+		var childWhereExpr sqlparser.Expr = compExpr
+		var childUpdateExprs sqlparser.UpdateExprs
+
+		switch fk.OnUpdate {
+		case sqlparser.Cascade:
+			// For the CASCADE type constraint, the update expressions are the same as the update expressions in the parent update query.
+			// We need to replace the column names in the update expressions with the corresponding child column names.
+			for _, updateExpr := range updClone.Exprs {
+				colIdx := fk.ParentColumns.FindColumn(updateExpr.Name.Name)
+				if colIdx >= 0 {
+					childUpdateExprs = append(childUpdateExprs, &sqlparser.UpdateExpr{
```

Comment:
I have a really hard time following code that is this deeply indented. don't you think that extracting a couple of methods here would help with readability? this method is too long and deep

Reply:
done